src/server/live-reload/watcher.js

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+
+var config = require('../config');
+var fs = require('fs');
+var log = require('../log');
+var ncp = require('ncp');
+var path = require('path');
+var prepare = require('../prepare');
+var Q = require('q');
+
+var eventIgnoreDuration = 150;
+var wwwRoot = 'www';
+var mergesRoot = 'merges';
+var unknownFileId = '__sim-unknown__';
+
+function Watcher(cssCb, nonCssCb) {
+    this.cssChangedCallback = cssCb;
+    this.nonCssChangedCallback = nonCssCb;
+    this.ignoreEvents = {};
+    this.mergesOverridePath = path.join(config.projectRoot, mergesRoot, config.platform);
+    this.mergesOverrideExists = fs.existsSync(this.mergesOverridePath);
+}
+
+Watcher.prototype.startWatching = function () {
+    this.wwwWatcher = fs.watch(path.join(config.projectRoot, wwwRoot), { recursive: true }, handleWwwWatcherEvent.bind(this));
+
+    if (this.mergesOverrideExists) {
+        this.mergesWatcher = fs.watch(path.join(config.projectRoot, mergesRoot, config.platform), { recursive: true }, handleMergesWatcherEvent.bind(this));
+    }
+};
+
+Watcher.prototype.stopWatching = function () {
+    if (this.wwwWatcher) {
+        this.wwwWatcher.close();
+        this.wwwWatcher = null;
+    }
+
+    if (this.mergesWatcher) {
+        this.mergesWatcher.close();
+        this.mergesWatcher = null;
+    }
+}
+
+function handleWwwWatcherEvent(event, fileRelativePath) {
+    handleWatcherEvent.bind(this)(wwwRoot, fileRelativePath);
+}
+
+function handleMergesWatcherEvent(event, fileRelativePath) {
+    handleWatcherEvent.bind(this)(mergesRoot, fileRelativePath);
+}
+
+function handleWatcherEvent(root, fileRelativePath) {
+    // fs.watch() will often send events more than once for the same modification, especially on Windows. A workaround is to block events generated by the same file for a short duration.
+    var ignoreId = fileRelativePath || unknownFileId;
+    
+    if (this.ignoreEvents[ignoreId]) {
+        return;
+    }
+
+    this.ignoreEvents[ignoreId] = true;
+    setTimeout(function () {
+        this.ignoreEvents[ignoreId] = false;
+    }.bind(this), eventIgnoreDuration);
+
+    if (!fileRelativePath) {
+        // fs.watch() doesn't always set the fileRelativePath argument properly. If that happens, let the user know.
+        log.warning('Could not reload the modified file because fs.watch() didn\'t specify which file was changed');
+
+        return;
+    }
+
+    // Make sure the modified file is either JS, HTML or CSS.
+    if (!fileIsOfType(fileRelativePath, ['.html', '.css', '.js'])) {
```

Comment:
I've changed the way live reload works. We no longer enforce specific file types (we listen for all changes). Also, for "resources", i.e. anything that is referenced in the html via a url, href or src attribute, we support refreshing without reloading the page. This includes but isn't limited to CSS and images.